sweep/sweeper.go

```diff
@@ -510,6 +510,82 @@ func (s *UtxoSweeper) feeRateForPreference(
 	return feeRate, nil
 }
 
+// removeLastSweepDescendants removes any transactions from the wallet that
+// spend outputs produced by the passed spendingTx. This needs to be done in
+// cases where we're not the only ones that can sweep an output, but there may
+// exists unconfirmed spends that spend outputs created by a sweep transaction.
+// The most common case for this is when someone sweeps our anchor outputs
+// after 16 blocks.
+func (s *UtxoSweeper) removeLastSweepDescendants(spendingTx *wire.MsgTx) error {
+	// Obtain all the past sweeps that we've done so far. We'll need these
+	// to ensure that if the spendingTx spends any of the same inputs, then
+	// we remove any transaction that may be spending those inputs from the
+	// wallet.
+	//
+	// TODO(roasbeef): can be last sweep here if we remove anything confirmed
+	// from the store?
+	pastSweepHashes, err := s.cfg.Store.ListSweeps()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Attempting to remove descendant txns invalidated by "+
+		"(txid=%v): %v", spendingTx.TxHash(), spew.Sdump(spendingTx))
+
+	// Construct a map of the inputs this transaction spends for each look
+	// up.
+	inputsSpent := make(map[wire.OutPoint]struct{})
+	for _, txIn := range spendingTx.TxIn {
+		inputsSpent[txIn.PreviousOutPoint] = struct{}{}
+	}
+
+	// We'll now go through each past transaction we published during this
+	// epoch and cross reference the spent inputs. If there're any inputs
+	// in common with the inputs the spendingTx spent, then we'll remove
+	// those.
+	//
+	// TODO(roasbeef): need to start to remove all transaction hashes after
+	// every N blocks (assumed point of no return)
+	for _, sweepHash := range pastSweepHashes {
+		sweepTx, err := s.cfg.Wallet.FetchTx(sweepHash)
+		if err != nil {
+			return err
+		}
+
+		// Transaction wasn't found in the wallet, may have already
+		// been
+		// replaced/removed.
+		if sweepTx == nil {
+			continue
+		}
+
+		// Check to see if this past sweep transaction spent any of the
+		// same inputs as spendingTx.
+		var isConflicting bool
+		for _, txIn := range sweepTx.TxIn {
```

Comment:
An alternative approach could be, instead of iterating all the sweep txes, we could instead check each of the outpoints spent by the sweep tx (the `inputsSpent`), to see if it's spent by another tx. something like,
```go
for _, txIn := range spendingTx.TxIn {
        if isSpent(txIn.PreviousOutPoint) {
                isConflicting = true
        }
}
```
Don't know if it's feasible to implement `isSpent`. If does, depending on the ordering and size of `pastSweepHashes`, we may gain some speed here by avoiding querying the db less.

Reply:
`isSpent` here can be implemented efficiently for full nodes (we just hit the UTXO set via RPC query), but not for neutrino nodes, as there's no sort of UTXO commitment in Bitcoin today. I think in practice, this loop should be rare-ish (tho it seems there're bots today that try to sweep anchor outputs when they see them). As mentioned above, we _should_ look into pruning this sweeper state, and it would actually be a relatively small jump to just start to remove the ones we know are no longer valid here.

Reply:
> but not for neutrino nodes, as there's no sort of UTXO commitment in Bitcoin today.

Yeah right, always forgot about light nodes.